### ISSUE TITLE
`DateFormat` time zone is not restored after parsing, affecting subsequent serialization

### ISSUE THREAD

**Issue by Marcono1234** [google/gson#2547]
# Gson version
2.10.1

# Java / Android version
Java 17

# Description
When deserializing a `java.util.Date` (and probably any of its subtypes) using a pattern with time zone, subsequent serialization using the same pattern is affected by it and uses the same time zone which was provided during deserialization. This can lead to undesired and unexpected behavior (e.g. one user can influence the date format of another user).

The underlying issue is that `DateFormat.parse` / `SimpleDateFormat.parse` can change the time zone of the instance, but does not restore it afterwards, see [JDK-6609675](https://bugs.openjdk.org/browse/JDK-6609675). So before any `parse` call it is necessary to obtain the time zone from the date format, then call `parse` and in a `finally` block restore the time zone again.
Based on the documentation it seems only the time zone is affected by this.

## Expected behavior
A call to `Gson.fromJson` should not affect subsequent `Gson.toJson` calls.

## Actual behavior
A call to `Gson.fromJson` can affect the output of subsequent `Gson.toJson` calls in case a date format which includes the time zone in the pattern is used.

# Reproduction steps
```java
TimeZone.setDefault(TimeZone.getTimeZone("UTC"));
Gson gson = new GsonBuilder().setDateFormat("yyyy-MM-dd HH:mm z").create();
Date date = new Date(0);

// As expected: 1970-01-01 00:00 UTC
System.out.println(gson.toJson(date));

gson.fromJson("\"1970-01-01 00:00 PST\"", Date.class);

// Unexpected: 1969-12-31 16:00 PST
System.out.println(gson.toJson(date));
```

**Comment by chrustkiran** [comment]
shall I check this issue?

**Comment by Marcono1234** [comment]
Do you mean whether you can submit a pull request fixing this bug?

No sorry, someone else is already working on a fix for this in #2549. There is probably no point in two people concurrently trying to fix this. But thank you anyway for the offer!
